cmd/kops/toolbox_instance_selector.go

```diff
@@ -0,0 +1,487 @@
+	flexible         = "flexible"
+)
+
+// Control Flag Constants
+const (
+	instanceGroupCount = "instance-group-count"
+	nodeCountMin       = "node-count-min"
+	nodeCountMax       = "node-count-max"
+	nodeVolumeSize     = "node-volume-size"
+	nodeSecurityGroups = "node-security-groups"
+	clusterAutoscaler  = "cluster-autoscaler"
+	usageClassSpot     = "spot"
+	usageClassOndemand = "on-demand"
+	igName             = "instance-group-name"
+	dryRun             = "dry-run"
+	output             = "output"
+)
+
+const (
+	nameRegex = `^[a-zA-Z0-9\-_]{1,128}$`
+)
+
+// InstanceSelectorOptions is a struct representing non-filter flags passed into instance-selector
+type InstanceSelectorOptions struct {
+	NodeCountMin       int32
+	NodeCountMax       int32
+	NodeVolumeSize     *int32
+	NodeSecurityGroups []string
+	ClusterAutoscaler  bool
+	InstanceGroupName  string
+	InstanceGroupCount int
+	Output             string
+	DryRun             bool
+}
+
+var (
+	toolboxInstanceSelectorLong = templates.LongDesc(i18n.T(`
+	Generate AWS EC2 on-demand or spot instance-groups by providing resource specs like vcpus and memory rather than instance types.`))
+
+	toolboxInstanceSelectorExample = templates.Examples(i18n.T(`
+
+	## Create a best-practices spot instance-group using a MixInstancesPolicy and Capacity-Optimized spot allocation strategy
+	## --flexible defaults to a 1:2 vcpus to memory ratio and 4 vcpus
+	kops toolbox instance-selector --usage-class spot --flexible --instance-group-name my-spot-mig
+
+	## Create a best-practices on-demand instance-group with custom vcpus and memory range filters
+	kops toolbox instance-selector --instance-group-name ondemand-ig --vcpus-min=2 --vcpus-max=4 --memory-min 2048 --memory-max 4096
+	`))
+
+	toolboxInstanceSelectorShort = i18n.T(`Generate on-demand or spot instance-group specs by providing resource specs like vcpus and memory.`)
+)
+
+// NewCmdToolboxInstanceSelector defines the cobra command for the instance-selector tool
+func NewCmdToolboxInstanceSelector(f *util.Factory, out io.Writer) *cobra.Command {
+	commandline := cli.New(
+		"instance-selector",
+		toolboxInstanceSelectorShort,
+		toolboxInstanceSelectorLong,
+		toolboxInstanceSelectorExample,
+		func(cmd *cobra.Command, args []string) {},
+	)
+	commandline.Command.Run = func(cmd *cobra.Command, args []string) {
+		ctx := context.TODO()
+
+		if err := rootCommand.ProcessArgs(args); err != nil {
+			exitWithError(err)
+		}
+		err := RunToolboxInstanceSelector(ctx, f, out, rootCommand.ClusterName(), &commandline)
+		if err != nil {
+			exitWithError(err)
+		}
+	}
+
+	cpuArchs := []string{"x86_64", "i386", "arm64"}
+	cpuArchDefault := "x86_64"
+	placementGroupStrategies := []string{"cluster", "partition", "spread"}
+	usageClasses := []string{usageClassSpot, usageClassOndemand}
+	usageClassDefault := usageClassOndemand
+	outputDefault := "yaml"
+	dryRunDefault := false
+	clusterAutoscalerDefault := false
+	nodeCountMinDefault := 2
+	nodeCountMaxDefault := 15
```

Comment:
I don't have a strong opinion about this but maybe the min and max should have the same defaults as `kops create ig`

Reply:
The intention of these defaults were to make cluster-autoscaler "just work™️". The min at 2 matches the default of `create ig` and max 15 allows CA to scale to a reasonable number.  I'm not opposed to matching min and max to 2 like `create ig` if there's good reason to.